addons/mass_mailing/controllers/main.py

```diff
@@ -18,7 +18,7 @@ def _check_mailing_email_token(self, mailing_id, res_id, email, token):
         if not (mailing_id and res_id and email and token):
             return False
         mailing = request.env['mailing.mailing'].sudo().browse(mailing_id)
-        return consteq(mailing._unsubscribe_token(res_id, email), token)
+        return consteq(mailing._generate_mailing_recipient_hash(res_id, email), token)
```

Comment:
I would keep the "token" in the name instead of "hash" (it's more a token than a hash because it depends on a secret, and not only on the data)